Replace AvalDimSharding, MeshDimAssignment and ShardedDeviceArray with the C++ object.

jblespiau opened the pull request:
The sharding and mesh_mapping fields are lists (instead of the current tuple), but they are immutable (I mean that if you modify it, it won't be reflected on the C++ object. But because it's a list you can decide to modify it).

It's internal to JAX, so it limits the issues with the possible confusion of having lists and not tuples, but it massively simplify the C++ as the casts are trivial.

This follows https://github.com/google/jax/pull/5634 and should be submitted before after.

jblespiau committed:
Replace AvalDimSharding and MeshDimAssignment with the C++ object.

The sharding and mesh_mapping fields are lists (instead of the current tuple), but they are immutable (I mean that if you modify it, it won't be reflected on the C++ object. But because it's a list you can decide to modify it).

It's internal to JAX, so it limits the issues with the possible confusion of having lists and not tuples, but it massively simplify the C++ as the casts are trivial.

2 new `# type: ignore` are needed, because mypy does not understand that List[A] is a List[Union[A, B]]:

```
jax/interpreters/pxla.py:1132: error: Argument "mesh_mapping" to "ShardingSpec" has incompatible type "List[Replicated]"; expected "List[Union[ShardedAxis, Replicated]]"  [arg-type]
jax/interpreters/pxla.py:1132: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
```

jblespiau commented:
NOTE:

I still need to silence mypy for some lines:
```
jax/interpreters/pxla.py:1141: error: Argument "mesh_mapping" to "ShardingSpec" has incompatible type "List[Replicated]"; expected "List[Union[ShardedAxis, Replicated]]"  [arg-type]
jax/interpreters/pxla.py:1141: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
jax/interpreters/pxla.py:1141: note: Consider using "Sequence" instead, which is covariant
```

I do not want to use `Sequence`, as this would be incorrect, the type *must* be a List. It just does not know Union merging.

apaszke commented on jax/interpreters/pxla.py:
```diff
@@ -479,9 +480,9 @@ def __init__(self,
-    self.sharding_spec = sharding_spec
+
+    pmap_lib.ShardedDeviceArray.__init__(self, aval, sharding_spec,
+                                         device_buffers)
```
Also, `super().__init__`?

jblespiau replied:
I renamed the CL to say it is also changing ShardedDEviceArray, which now extends a C++ type.

It's done by:
- extending the C++ object.
- it will serve aval, sharding_spec and device_buffers, which are removed from __slots__
- calling this contructor.

apaszke reviewed (commented):
LGTM once the comments are addressed. Why do you have to be so careful about passing in lists? I thought we're using pybind to do all the conversions, and it should successfully convert tuples to `std::vector`s

jblespiau commented on jax/interpreters/pxla.py:
```diff
@@ -104,31 +104,20 @@ class Replicated(NamedTuple):
-    self.mesh_mapping = tuple(mesh_mapping)
+# We use a dummy if-statement to reduce the noise in the PR and confirm that
+# nothing changes.
+# TODO(jblespiau): Remove in its own CL to reduce noise.
```
If you mean this constant, I removed it.

jblespiau replied:
(In practice, as you will see, it's doing a very poor job at comparing files, and everything appears to be different^^

jblespiau commented:
> LGTM once the comments are addressed. Why do you have to be so careful about passing in lists? I thought we're using pybind to do all the conversions, and it should successfully convert tuples to `std::vector`s

Pybind11 converts py::list (i.e. Python lists) into std::vector, and the other way around.

Tuples (py::tuple) are not convertible to std::vector (I imagine they are to std::tuple).

So if you pass a tuple to a function that expects a std::vector in C++, it will fail.

jblespiau commented on jax/interpreters/pxla.py:
```diff
@@ -1106,33 +1107,38 @@ def _pmap_sharding_spec(nrep, axis_size, npart, parts, sharded_aval, map_axis: O
     return ShardingSpec(
         sharding=[_UNSHARDED_INSTANCE] * len(aval.shape),
-        mesh_mapping=maybe_replicate)
+        mesh_mapping=maybe_replicate)  # type: ignore
```
See the comment at the top of the PR:

```
NOTE:

I still need to silence mypy for some lines:

jax/interpreters/pxla.py:1141: error: Argument "mesh_mapping" to "ShardingSpec" has incompatible type "List[Replicated]"; expected "List[Union[ShardedAxis, Replicated]]"  [arg-type]
jax/interpreters/pxla.py:1141: note: "List" is invariant -- see http://mypy.readthedocs.io/en/latest/common_issues.html#variance
jax/interpreters/pxla.py:1141: note: Consider using "Sequence" instead, which is covariant
```

jblespiau commented on jax/interpreters/pxla.py:
```diff
@@ -1106,33 +1107,38 @@ def _pmap_sharding_spec(nrep, axis_size, npart, parts, sharded_aval, map_axis: O
-      sharding=pspec.sharding,
-      mesh_mapping=(Replicated(axis_size),) + maybe_replicate + pspec.mesh_mapping)
+        sharding=pspec.sharding,
+        mesh_mapping=[Replicated(axis_size)] + maybe_replicate + pspec.mesh_mapping)  # type: ignore
```
Same reason as above. mypy does not understand that a List[Replicated] is valid for a List[Union[Replicated, ShardedAxis].

jblespiau commented on jax/interpreters/pxla.py:
```diff
@@ -1106,33 +1107,38 @@ def _pmap_sharding_spec(nrep, axis_size, npart, parts, sharded_aval, map_axis: O
+    idx = map_axis
+    new_sharding = l[:idx] + [Unstacked(axis_size)] + l[idx:]
     return ShardingSpec(
-      sharding=tuple_insert(pspec.sharding, map_axis, Unstacked(axis_size)),
```
tuple_insert does not work on lists, because it concanates a 1-element tuple:

def tuple_insert(t, idx, val):
  assert 0 <= idx <= len(t), (idx, len(t))
  return t[:idx] + (val,) + t[idx:] <<< This is invalid to concatenate a list and a tuple.

jblespiau commented on jax/interpreters/pxla.py:
```diff
@@ -270,6 +259,14 @@ def __hash__(self):
+  setattr(pmap_lib.ShardingSpec, "indices", indices)
+  setattr(pmap_lib.ShardingSpec, "__eq__", __eq__)
+  setattr(pmap_lib.ShardingSpec, "__hash__", __hash__)
+  setattr(pmap_lib.ShardingSpec, "__repr__", __repr__)
```
Done.

jblespiau commented:
Updated the CL.

Removed formatting and set the attribute directly on `ShardingSpec`.

For the typing errors, the tuple <-> st::vector conversion, tuple_insert, I explained why I opted for this option (typing errors do occur with mypy, std::vectors are not convertible to tuples, tuple_insert does not work on lists).

jblespiau commented:
Closing, I will submit this internally to prevent breakages.

jblespiau commented:
It's being reviewed internally as cr/352187814.
